File: bindings/python/tests/cassette_replay.py
"""Cassette replay server for Python tests.

Reads real cassette JSON files from aimux-providers/tests/cassettes/
and serves them via a local HTTP server with the same matching logic
as the Rust `common::replay` module.

The server runs in a separate process (multiprocessing) because PyO3's
block_on blocks the calling thread.
"""

import json
import os
import socket
import time
from http.server import HTTPServer, BaseHTTPRequestHandler
from multiprocessing import Process
from pathlib import Path

CASSETTE_BASE = Path(__file__).resolve().parents[3] / "aimux-providers" / "tests" / "cassettes"


def _find_free_port():
    with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
        s.bind(("127.0.0.1", 0))
        return s.getsockname()[1]


def _load_cassettes(provider):
    """Load all cassette JSON files for a provider."""
    cdir = CASSETTE_BASE / provider
    cassettes = []
    for f in sorted(cdir.glob("*.json")):
        try:
            raw = json.loads(f.read_text())
            body = raw.get("request", {}).get("body", {})
            if isinstance(body, str):
                try:
                    body = json.loads(body)
                except Exception:
                    body = {}
            cassettes.append({
                "scenario": raw.get("scenario", f.name),
                "req_path": raw.get("request", {}).get("path", "/"),
                "req_method": raw.get("request", {}).get("method", "POST"),
                "req_body": body if isinstance(body, dict) else {},
                "resp_status": raw.get("response", {}).get("status", 200),
                "resp_headers": raw.get("response", {}).get("headers", {}),
                "resp_body": raw.get("response", {}).get("body", ""),
            })
        except Exception:
            pass
    return cassettes


def _score(cassette_body, req_body):
    """Score a cassette against the incoming request body."""
    if not isinstance(cassette_body, dict) or not isinstance(req_body, dict):
        return 0
    s = 0
    for key, val in cassette_body.items():
        if not isinstance(val, (str, bool, int, float)):
            continue
        if req_body.get(key) == val:
            if key == "model":
                s += 100
            elif key == "stream":
                s += 10
            else:
                s += 1
    return s


def _find_best_match(cassettes, method, path, req_body):
    """Find the best matching cassette for a request."""
    group = [c for c in cassettes if c["req_method"] == method and c["req_path"] == path]
    if not group:
        return None
    best = group[0]
    best_score = _score(best["req_body"], req_body)
    for c in group[1:]:
        s = _score(c["req_body"], req_body)
        if s > best_score:
            best_score = s
            best = c
    return best


def _server_proc(port, cassettes):
    """Mock HTTP server process."""

    class Handler(BaseHTTPRequestHandler):
        def do_POST(self):
            body = self.rfile.read(int(self.headers.get("Content-Length", 0)))
            try:
                req_body = json.loads(body)
            except Exception:
                req_body = {}

            match = _find_best_match(cassettes, "POST", self.path, req_body)
            if not match:
                self.send_response(404)
                self.send_header("content-type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps({"error": f"no cassette for POST {self.path}"}).encode())
                return

            resp = match["resp_body"].encode() if isinstance(match["resp_body"], str) else match["resp_body"]
            self.send_response(match["resp_status"])
            for k, v in match["resp_headers"].items():
                self.send_header(k, v)
            self.send_header("content-length", str(len(resp)))
            self.end_headers()
            self.wfile.write(resp)

        def do_GET(self):
            match = _find_best_match(cassettes, "GET", self.path, {})
            if not match:
                self.send_response(404)
                self.send_header("content-type", "application/json")
                self.end_headers()
                self.wfile.write(json.dumps({"error": f"no cassette for GET {self.path}"}).encode())
                return
            resp = match["resp_body"].encode() if isinstance(match["resp_body"], str) else match["resp_body"]
            self.send_response(match["resp_status"])
            for k, v in match["resp_headers"].items():
                self.send_header(k, v)
            self.send_header("content-length", str(len(resp)))
            self.end_headers()
            self.wfile.write(resp)

        def log_message(self, *args):
            pass

    server = HTTPServer(("127.0.0.1", port), Handler)
    server.serve_forever()


class CassetteServer:
    """Context manager that starts a cassette replay server in a subprocess."""

    def __init__(self, provider):
        self.cassettes = _load_cassettes(provider)
        if not self.cassettes:
            raise RuntimeError(f"No cassettes found for provider: {provider}")
        self.port = _find_free_port()
        self.url = f"http://127.0.0.1:{self.port}"
        self.proc = Process(target=_server_proc, args=(self.port, self.cassettes))

    @property
    def count(self):
        return len(self.cassettes)

    def __enter__(self):
        self.proc.start()
        # Wait for server to be ready
        for _ in range(50):
            try:
                with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
                    s.connect(("127.0.0.1", self.port))
                    break
            except ConnectionRefusedError:
                time.sleep(0.05)
        return self

    def __exit__(self, *args):
        self.proc.terminate()
        self.proc.join(timeout=2)

File: bindings/node/__test__/cassette-replay.ts
// cassette-replay.ts — Lightweight cassette replay server for Node tests.
//
// Reads real cassette JSON files (from aimux-providers/tests/cassettes/),
// serves them via a local HTTP server with the same matching logic as
// the Rust `common::replay` module: group by (method, path), score by
// model + stream flag, fall back to first cassette.

import { createServer, type Server, type IncomingMessage, type ServerResponse } from 'node:http'
import { readFileSync, readdirSync } from 'node:fs'
import { join, dirname } from 'node:path'
import { fileURLToPath } from 'node:url'

const __dirname = dirname(fileURLToPath(import.meta.url))
const CASSETTE_BASE = join(__dirname, '..', '..', '..', 'aimux-providers', 'tests', 'cassettes')

interface Cassette {
  scenario: string
  reqPath: string
  reqMethod: string
  reqBody: any
  respStatus: number
  respHeaders: Record<string, string>
  respBody: string
}

function loadCassettes(provider: string): Cassette[] {
  const dir = join(CASSETTE_BASE, provider)
  const files = readdirSync(dir).filter((f) => f.endsWith('.json'))
  const cassettes: Cassette[] = []
  for (const file of files) {
    try {
      const raw = JSON.parse(readFileSync(join(dir, file), 'utf-8'))
      const body = raw.request?.body
      // body can be a JSON object or a string
      let reqBody = body
      if (typeof body === 'string') {
        try { reqBody = JSON.parse(body) } catch { reqBody = null }
      }
      cassettes.push({
        scenario: raw.scenario || file,
        reqPath: raw.request?.path || '/',
        reqMethod: raw.request?.method || 'POST',
        reqBody: reqBody || {},
        respStatus: raw.response?.status || 200,
        respHeaders: raw.response?.headers || {},
        respBody: raw.response?.body || '',
      })
    } catch (e) {
      // skip malformed cassettes
    }
  }
  return cassettes
}

// Score a cassette against the incoming request body (same logic as Rust replay)
function score(cassetteBody: any, reqBody: any): number {
  if (typeof cassetteBody !== 'object' || cassetteBody === null) return 0
  if (typeof reqBody !== 'object' || reqBody === null) return 0
  let s = 0
  for (const [key, val] of Object.entries(cassetteBody)) {
    if (typeof val !== 'string' && typeof val !== 'boolean' && typeof val !== 'number') continue
    if (reqBody[key] === val) {
      if (key === 'model') s += 100
      else if (key === 'stream') s += 10
      else s += 1
    }
  }
  return s
}

function findBestMatch(cassettes: Cassette[], method: string, path: string, reqBody: any): Cassette | null {
  // Group by (method, path)
  const group = cassettes.filter((c) => c.reqMethod === method && c.reqPath === path)
  if (group.length === 0) return null

  // Score and pick best
  let best = group[0]
  let bestScore = score(best.reqBody, reqBody)
  for (let i = 1; i < group.length; i++) {
    const s = score(group[i].reqBody, reqBody)
    if (s > bestScore) {
      bestScore = s
      best = group[i]
    }
  }
  return best
}

export class CassetteServer {
  private server: Server
  private cassettes: Cassette[]
  public url: string

  constructor(provider: string) {
    this.cassettes = loadCassettes(provider)
    if (this.cassettes.length === 0) {
      throw new Error(`No cassettes found for provider: ${provider}`)
    }

    this.server = createServer((req, res) => this.handle(req, res))
    this.url = '' // set in start()
  }

  private handle(req: IncomingMessage, res: ServerResponse) {
    let body = ''
    req.on('data', (chunk) => (body += chunk))
    req.on('end', () => {
      let reqBody: any = null
      try { reqBody = JSON.parse(body) } catch {}

      const match = findBestMatch(this.cassettes, req.method || 'POST', req.url || '/', reqBody)
      if (!match) {
        res.writeHead(404, { 'content-type': 'application/json' })
        res.end(JSON.stringify({ error: `no cassette for ${req.method} ${req.url}` }))
        return
      }

      res.writeHead(match.respStatus, match.respHeaders)
      res.end(match.respBody)
    })
  }

  async start(): Promise<void> {
    return new Promise((resolve) => {
      this.server.listen(0, '127.0.0.1', () => {
        const addr = this.server.address() as any
        this.url = `http://127.0.0.1:${addr.port}`
        resolve()
      })
    })
  }

  async stop(): Promise<void> {
    return new Promise((resolve) => this.server.close(() => resolve()))
  }

  get count(): number {
    return this.cassettes.length
  }
}
